### DISCUSSION TITLE
unsure about interaction between exec() and shortcuts

### DISCUSSION THREAD

**Discussion by cfritze64** [johnlindquist/kit#499] [Q&A]
John and community:

I have some automations in Apple's Shortcuts app, especially now what it's available in Monterey on MacOS.  I'm building a little script that is a front-end picker for running Shortcuts or opening them for editing in the Shortcuts app. I've hit a snag. Eventually in my script, I'm trying a line like
```
exec(`shortcuts run "${aShortcutsName}"`)
```

That fails. It works if I use the terminal function instead of exec, but that pops up a stray  terminal window that I don't need.  So am I misunderstanding the exec function (I'm a programming tyro), or is there an interaction happening between the shortcuts terminal command (probably new in Monterey) and the scriptkit shortcut global?

**Comment by johnlindquist** [comment]
Is it failing with an error? Or just silently?

I'm curious if it might be related to this: https://github.com/johnlindquist/kit/issues/487

I need to investigate...

**Comment by cfritze64** [comment]
It's weird. Sometimes it fails silently, sometimes I get a "Processing" message in scriptkit's window and a blue spinny ring. The latter seems forever, and I have to quit scriptkit ungracefully to get my machine back. 

Another wrinkle: the simple exec command above fails, but this one works:
```
exec(`shortcuts list -f ${theFolder}`)
```

**Reply by johnlindquist** [reply]
Can you try using node's exec:

Just put this at the top of your script, it should override the shelljs exec:

import { exec } from 'child_process'

**Reply by cfritze64** [reply]
Wild. If I use the node's exec as you suggest, this one stops working (fails silently):
```
exec(`shortcuts list -f ${theFolder}`) ;
```

but
```
terminal(`shortcuts list -f ${theFolder}`) ;
```
still works (of course)

**Reply by johnlindquist** [reply]
Oh, try adding a callback to exec:
 (pasted from the shelljs docs, I'm  on my phone)


```js
var child = exec('some_long_running_process', {async:true});
child.stdout.on('data', function(data) {
  /* ... do something with data ... */
});
```

I'm assuming because your exec doesn't have callback, the app considers the script complete and exits before the command can run.

You could also try an:

```js
await wait(5000)
```

After your exec.

Or use 
```js
await $`shortcuts run foo`
```

**Reply by cfritze64** [reply]
Oh man. I got all turned around the axle playing with callbacks and async/awaits in trying to make exec work. That's the fault of my inexperience. Finally I stumbled onto node's execSync. Now I can run shell commands that return output like:
```
  const output = execSync(`shortcuts list -f ${theFolder}`, { encoding: 'utf-8'});
```

and launch commands where I don't collect a returned value like:
```
  execSync(`shortcuts run "${theSc}"`);
```

**Reply by johnlindquist** [reply]
Awesome. Glad you got it sorted. 

I'll add some of this stuff to the docs/guide. It seems to be something a lot of people stumble over
